src/coprocessor/cache.rs

```diff
@@ -0,0 +1,442 @@
+                region.version += 1;
+                let mut keys: Vec<DistSQLCacheKey> = Vec::new();
+                for (key, _) in (&region.cached_items).iter() {
+                    keys.push(key.to_string());
+                }
+                Some(keys)
+            }
+        };
+        match keys {
+            None => {
+                let entry = RegionDistSQLCacheEntry {
+                    version: 1,
+                    cached_items: HashMap::new(),
+                };
+                self.regions.insert(region_id, entry);
+            }
+            Some(keys) => for i in keys {
+                self.remove(&i);
+            },
+        };
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    pub fn evict_region_and_enable(&mut self, region_id: u64) {
+        self.evict_region(region_id);
+        self.enable_cache = true;
+    }
+
+    pub fn enable(&mut self) {
+        self.enable_cache = true;
+    }
+
+    pub fn disable(&mut self) {
+        self.enable_cache = false;
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.max_size
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.len() == 0
+    }
+
+    pub fn update_capacity(&mut self, capacity: usize) {
+        self.max_size = capacity;
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+    }
+
+    fn update_regions(&mut self, region_id: u64, k: DistSQLCacheKey) {
+        let opt = match self.regions.get_mut(&region_id) {
+            Some(entry) => {
+                entry.cached_items.insert(k, 1);
+                None
+            }
+            None => {
+                let mut rmap = HashMap::new();
+                rmap.insert(k, 1);
+                Some(rmap)
+            }
+        };
+        if let Some(rmap) = opt {
+            let entry = RegionDistSQLCacheEntry {
+                version: 0,
+                cached_items: rmap,
+            };
+            self.regions.insert(region_id, entry);
+        }
+    }
+
+    #[inline]
+    fn remove_lru(&mut self) {
+        match self.map.pop_front() {
+            None => (),
+            Some((_, entry)) => {
+                self.size -= entry.size;
+            }
+        };
+    }
+}
+
+// DistSQL cache size unit is byte, default is 256MB
+pub const DEFAULT_DISTSQL_CACHE_SIZE: usize = 256 * 1024 * 1024;
+// DistSQL cache entry max size unit is byte, default is 5MB
+pub const DEFAULT_DISTSQL_CACHE_ENTRY_MAX_SIZE: usize = 5 * 1204 * 1024;
+
+lazy_static! {
+    pub static ref DISTSQL_CACHE: Arc<Mutex<DistSQLCache>> =
```

Comment:
Using RwLock may be better here?

Reply:
Seems only
```
version = DISTSQL_CACHE.lock().unwrap().get_region_version(region_id);
```
can benefit from RwLock. What about combine get_region_version and get method into one function call?

Reply:
sorry for the later response, I concern the lock contention here if the operation is frequent. If the get cache is fast, we can use a spinlock instead later.

Reply:
@siddontang sure. I think spinlock should be better than switch thread context out.

Reply:
BTW, `Arc` is not need here.